dhall/src/Dhall/Parser/Combinators.hs

```diff
@@ -69,11 +74,42 @@ laxSrcEq (Src p q _) (Src p' q' _) = eq p p' && eq q q'
         a == a' && b == b'
 {-# INLINE laxSrcEq #-}
 
+-- | A way to control how comments get parsed
+--
+-- Parsing comment as whitespace is fast and convenient when we don't care
+-- about the comments. We often need to keep comments thoughout
+-- transformations.
+data CommentControl
+  = CommentIsWhitespace
+  | CommentIsNeeded
```

Comment:
I think I would suggest these names and documentation:

```suggestion
data WhitespaceControl
  = UnsupportedCommentsForbidden
  -- ^ The @whitespace@ parser rejects comments, meaning that comments need to be
  -- parsed explicitly in order to avoid parse errors.  We use this to detect comments that are
  -- not yet preserved and formatted.
  | UnsupportedCommentsPermitted
  -- ^ The @whitespace@ parser accepts comments, meaning that comments are accepted
  -- anywhere within the source code, even if they are not preserved or formatted.
```